plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/repository/M2RepositoryService.java

```diff
@@ -98,26 +136,63 @@ private Artifact resolveArtifact(MavenCoordinates mavenCoordinates) throws Libra
       } else {
         attributes.add(UpdateClasspathAttributeUtil.createNonDependencyAttribute());
       }
+      if (libraryFile.getSourceUri() != null) {
+        addUriAttribute(attributes, CLASSPATH_ATTRIBUTE_SOURCE_URL, libraryFile.getSourceUri());
+      }
       if (libraryFile.getJavadocUri() != null) {
-        attributes.add(JavaCore.newClasspathAttribute(IClasspathAttribute.JAVADOC_LOCATION_ATTRIBUTE_NAME,
-                                                      libraryFile.getJavadocUri().toURL().toString()));
+        addUriAttribute(attributes, IClasspathAttribute.JAVADOC_LOCATION_ATTRIBUTE_NAME, libraryFile.getJavadocUri());
       }
       return attributes.toArray(new IClasspathAttribute[0]);
-    } catch (CoreException | MalformedURLException ex) {
+    } catch (CoreException ex) {
       throw new LibraryRepositoryServiceException("Could not create classpath attributes", ex);
     }
   }
 
-  private IPath getSourceLocation(LibraryFile libraryFile) {
-    if (libraryFile.getSourceUri() == null) {
-      return getSourceJarLocation(libraryFile.getMavenCoordinates());
+  protected void addUriAttribute(List<IClasspathAttribute> attributes, String attributeName, URI uri) {
+    try {
+      attributes.add(JavaCore.newClasspathAttribute(attributeName, uri.toURL().toString()));
+    } catch (MalformedURLException | IllegalArgumentException ex) {
+      // disregard invalid URL
+    }
+  }
+
+  private IPath getSourceLocation(MavenCoordinates mavenCoordinates, URL sourceUrl) {
+    if (sourceUrl == null) {
+      return getMavenSourceJarLocation(mavenCoordinates);
     } else {
-      // download the file and return path to it
-      // TODO https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/800
-      return new Path("/downloaded/source/file"); //$NON-NLS-1$
+      return getDownloadedSourceLocation(mavenCoordinates, sourceUrl);
     }
   }
 
+  private IPath getDownloadedSourceLocation(MavenCoordinates mavenCoordinates, URL sourceUrl) {
+    
+    try {
+      IPath downloadFolder = getDownloadedFilesFolder(mavenCoordinates);
+      IPath path = new FileDownloader(downloadFolder).download(sourceUrl);
```

Comment:
I question whether we want to go down this path.  Source jars on MC have a predictable layout, but handling arbitrary source jars is fraught.